connection.go

```diff
@@ -785,12 +845,40 @@ func (conn *Connection) newFuture() (fut *Future) {
 	return
 }
 
+func (conn *Connection) contextWatchdog(fut *Future, ctx context.Context) {
+	select {
+	case <-fut.done:
+	default:
+		select {
+		case <-ctx.Done():
+			conn.cancelFuture(fut, fmt.Errorf("context is done"))
+		default:
```

Comment:
I don't get it.

Reply:
`conn.cancelFuture()` removes a Future object from the internal queue with setting an error.

Reply:
I don't get `default: default:` sequence logic.

Reply:
There are two points to use it:
1) If we have both channels "done" (context.Done() and future.done) the default choice of case branch in simple select is random, like:

```go
select {
case <-fut.done:
case <-req.context.Done():
}
```

The proposed logic makes this behavior defined.

2) This logic works faster due to internal implementation of select in go. You can read the [thread](https://groups.google.com/g/golang-dev/c/jX4oQEls3uk) for more details.

Reply:
Added.